addons/stock/models/stock_picking.py

```diff
@@ -830,6 +841,158 @@ def _create_backorder(self, backorder_moves=[]):
+        else:
+            raise UserError(_('Unknow stream.'))
+
+        for (parent, responsible), childs in child_object_by_parent_object:
+            childs = list(childs)
+            childs_recordset = self.env[childs[0]._name].browse([c.id for c in childs])
+            # Get the note
+            note = note_method(orig_obj_changes, childs_recordset, orig_obj_child_object_rel)
+
+            self.env['mail.activity'].create({
+                'activity_type_id': self.env.ref('mail.mail_activity_data_todo').id,
+                'note': note,
+                'user_id': responsible.id,
+                'res_id': parent.id,
+                'res_model_id': self.env['ir.model'].search([('model', '=', parent._name)], limit=1).id,
+            })
+
+    def _log_less_quantities_than_expected(self, moves):
+        """ param: moves a dict of moves and their new and old quantity
+        e.g : {'move_id': (4, 5), 'move_id2': (6, 7)}
+        """
+        def _keys_in_sorted(move):
+            return (move.picking_id.id, move.product_id.responsible_id.id)
+
+        def _keys_in_groupby(move):
+            return (move.picking_id, move.product_id.responsible_id)
+
+        def _generate_note_missing_quantity(quantity_by_move, move_dest_ids, moves_orig_dest_rel):
+            origin_moves = self.env['stock.move'].browse([m .id for m in moves_orig_dest_rel.values()])
+            origin_picking = origin_moves.mapped('picking_id')
+
+            note = '<div class="alert alert-warning">' + _('Exception(s) occurred on the picking\
+            <a href="#" data-oe-model=stock.picking data-oe-id=%d>%s</a>. \
+            Manual actions may be needed.') % (origin_picking.id, origin_picking.name) + '</div>'
+
+            # Add information about moves in note.
+            note += '<p>' + _('Quantity processed: ') + '</p><ul>'
+            for move in move_dest_ids:
+                (new_qty, old_qty) = quantity_by_move[moves_orig_dest_rel[move]]
+                # moves_rel contains the origin moves from the destination moves. It means that it return a moves passed as arguments and we take each element of the tuple for the quantity
+                note += '<li>' + _('%s %s of %s processed instead of %s') % (new_qty, move.product_uom.name, move.product_id.display_name, old_qty) + '</li>'
+            note += '</ul>'
+            # Find impacted pickings for a specific picking due to the move_dest_ids.
+            impacted_pickings = origin_picking._get_impacted_pickings(move_dest_ids) - move_dest_ids.mapped('picking_id')
+            # Add impacted pickings on the note.
+            if impacted_pickings:
+                note += '<p>' + _('Next transfer(s) impacted:') + '</p><ul>'
+                for pick in impacted_pickings:
+                    note += '<li><a href=# data-oe-model=stock.picking data-oe-id=%d>%s</a></li>' % (pick.id, pick.name)
+                note += '</ul>'
+            return note
```

Comment:
it would be cleaner to use a qweb view, no? and pass the arguments in the rendering context